gpio/src/gpio.rs

```diff
@@ -0,0 +1,943 @@
+pub(crate) const VIRTIO_GPIO_DIRECTION_OUT: u8 = 0x01;
+pub(crate) const VIRTIO_GPIO_DIRECTION_IN: u8 = 0x02;
+
+/// Virtio GPIO IRQ types
+pub(crate) const VIRTIO_GPIO_IRQ_TYPE_NONE: u16 = 0x00;
+pub(crate) const VIRTIO_GPIO_IRQ_TYPE_EDGE_RISING: u16 = 0x01;
+pub(crate) const VIRTIO_GPIO_IRQ_TYPE_EDGE_FALLING: u16 = 0x02;
+pub(crate) const VIRTIO_GPIO_IRQ_TYPE_EDGE_BOTH: u16 =
+    VIRTIO_GPIO_IRQ_TYPE_EDGE_RISING | VIRTIO_GPIO_IRQ_TYPE_EDGE_FALLING;
+pub(crate) const VIRTIO_GPIO_IRQ_TYPE_LEVEL_HIGH: u16 = 0x04;
+pub(crate) const VIRTIO_GPIO_IRQ_TYPE_LEVEL_LOW: u16 = 0x08;
+const VIRTIO_GPIO_IRQ_TYPE_ALL: u16 = VIRTIO_GPIO_IRQ_TYPE_EDGE_BOTH
+    | VIRTIO_GPIO_IRQ_TYPE_LEVEL_HIGH
+    | VIRTIO_GPIO_IRQ_TYPE_LEVEL_LOW;
+
+/// Virtio GPIO Configuration
+#[derive(Clone, Debug, PartialEq)]
+#[repr(C)]
+pub(crate) struct VirtioGpioConfig {
+    pub(crate) ngpio: Le16,
+    pub(crate) padding: Le16,
+    pub(crate) gpio_names_size: Le32,
+}
+
+/// Trait that represents an GPIO Device.
+///
+/// This trait is introduced for development purposes only, and should not
+/// be used outside of this crate. The purpose of this trait is to provide a
+/// mock implementation for the GPIO driver so that we can test the GPIO
+/// functionality without the need of a physical device.
+pub(crate) trait GpioDevice: Send + Sync + 'static {
+    fn open(device: u32) -> Result<Self>
+    where
+        Self: Sized;
+
+    fn get_num_lines(&self) -> Result<u16>;
+    fn get_line_name(&self, gpio: u16) -> Result<String>;
+    fn get_direction(&self, gpio: u16) -> Result<u8>;
+    fn set_direction(&self, gpio: u16, dir: u8, value: u32) -> Result<()>;
+    fn get_value(&self, gpio: u16) -> Result<u8>;
+    fn set_value(&self, gpio: u16, value: u32) -> Result<()>;
+
+    fn set_irq_type(&self, gpio: u16, value: u16) -> Result<()>;
+    fn wait_for_interrupt(&self, gpio: u16) -> Result<()>;
+}
+
+pub(crate) struct PhysDeviceState {
+    // See wait_for_interrupt() for explanation of Arc.
+    request: Option<Arc<LineRequest>>,
+    buffer: Option<EdgeEventBuffer>,
+}
+
+pub(crate) struct PhysDevice {
+    chip: Chip,
+    ngpio: u16,
+    state: Vec<RwLock<PhysDeviceState>>,
+}
+
+unsafe impl Send for PhysDevice {}
+unsafe impl Sync for PhysDevice {}
+
+impl GpioDevice for PhysDevice {
+    fn open(device: u32) -> Result<Self>
+    where
+        Self: Sized,
+    {
+        let path = format!("/dev/gpiochip{}", device);
+        let chip = Chip::open(&path).map_err(Error::GpiodFailed)?;
+        let ngpio = chip.get_num_lines() as u16;
+
+        // Can't set a vector to all None easily
+        let mut state: Vec<RwLock<PhysDeviceState>> = Vec::new();
+        state.resize_with(ngpio as usize, || {
+            RwLock::new(PhysDeviceState {
+                request: None,
+                buffer: None,
+            })
+        });
+
+        Ok(PhysDevice { chip, ngpio, state })
+    }
+
+    fn get_num_lines(&self) -> Result<u16> {
+        Ok(self.ngpio)
```

Comment:
I have reviewed this file twice now and both time I found the usage of variables named "gpio" very confusing.  Here the function is get_num_lines() but we are returning ngpio.  More like that below...

Reply:
I get what you are saying, the GPIO pins are called with pin, line, gpio. etc, in all code/documentation we can find. I have decided, now, to stick with how the virtio-spec names it, which is `gpio` (for a gpio pin/line) and `ngpio` (for their count). And so renamed this routine to `get_num_gpios()`.

Since the PhysDevice implementation depends on GpioLib, I have kept it as `line` there.